### PR TITLE
GST-plugin: Add new features

### PR DESCRIPTION
**Pull request by tianjunwork**
Signed-off-by: Jun Tian <jun.tian@intel.com>

### CODE REVIEW

File: gstreamer-plugin/gstsvthevcenc.c [hunk 1]
```diff
@@ -466,45 +494,44 @@ gst_svthevcenc_get_property (GObject * object, guint property_id,
...
+    case PROP_BASE_LAYER_SWITCH_MODE:
+        g_value_set_uint (value,
+          svthevcenc->svt_config->baseLayerSwitchMode);
       break;
     case PROP_PRED_STRUCTURE:
       g_value_set_uint (value, svthevcenc->svt_config->predStructure);
       break;
-    case PROP_GOP_SIZE:
+    case PROP_KEY_INT_MAX:
       g_value_set_int (value, svthevcenc->svt_config->intraPeriodLength + 1);
```
Comment (Yeongjin-Jeong) [inline gstreamer-plugin/gstsvthevcenc.c]: It need to handle intraPeriodLength < 0 (e.g. no intra, auto)

File: gstreamer-plugin/gstsvthevcenc.c [hunk 2]
```diff
@@ -731,19 +784,18 @@ set_default_svt_configuration (EB_H265_ENC_CONFIGURATION * svt_config)
...
   svt_config->compressedTenBitFormat = FALSE;
-  svt_config->profile = 1;
+  svt_config->profile = 2;
   svt_config->tier = 0;
   svt_config->level = 0;
-  svt_config->injectorFrameRate =
-      (PROP_SPEEDCONTROL_DEFAULT > 0) ? PROP_SPEEDCONTROL_DEFAULT : 60 << 16;
-  svt_config->speedControlFlag = (PROP_SPEEDCONTROL_DEFAULT > 0);
-  svt_config->latencyMode = PROP_LOWLATENCY_DEFAULT;
+  svt_config->injectorFrameRate = 60 << 16;
```
Comment (Yeongjin-Jeong) [inline gstreamer-plugin/gstsvthevcenc.c]: Does that require a shift operation? If so, I think comments should be added to the document and API header.

File: gstreamer-plugin/gstsvthevcenc.c [hunk 3]
```diff
@@ -731,19 +784,18 @@ set_default_svt_configuration (EB_H265_ENC_CONFIGURATION * svt_config)
   svt_config->encoderBitDepth = 8;
   svt_config->encoderColorFormat = EB_YUV420;
   svt_config->compressedTenBitFormat = FALSE;
-  svt_config->profile = 1;
+  svt_config->profile = 2;
```
Comment (Yeongjin-Jeong) [inline gstreamer-plugin/gstsvthevcenc.c]: If input profile is not 4 in Y444 case, bad parameter error returns from svt encoder.

File: gstreamer-plugin/gstsvthevcenc.c [hunk 4]
```diff
@@ -466,45 +494,44 @@ gst_svthevcenc_get_property (GObject * object, guint property_id,
...
       break;
+    case PROP_QP_MAX:
+      g_value_set_uint (value, svthevcenc->svt_config->maxQpAllowed);
+      break;
+    case PROP_QP_MIN:
+      g_value_set_uint (value, svthevcenc->svt_config->minQpAllowed);
+      break;
     case PROP_DEBLOCKING:
-      g_value_set_boolean (value, svthevcenc->svt_config->disableDlfFlag == 0);
+      g_value_set_boolean (value, svthevcenc->svt_config->disableDlfFlag == 1);
```
Comment (Yeongjin-Jeong) [inline gstreamer-plugin/gstsvthevcenc.c]: Is "PROP_DEBLOCKING set TRUE" correct when "disableDlfFlag == 0" ?

**Comment by Yeongjin-Jeong** [comment]
https://github.com/OpenVisualCloud/SVT-HEVC/pull/157  does not have some properties in here ("pred-struct", "deblocking", "sao", "constrained-intra"). Can I add them there?

File: gstreamer-plugin/gstsvthevcenc.c (hunk 2 above)
Comment (tianjunwork) [reply gstreamer-plugin/gstsvthevcenc.c]: There is a conversion of injectorFrameRate api internally. 
EbAppConfig.c, line 237:    
    cfg->injectorFrameRate = strtoul(value, NULL, 0);
    if (cfg->injectorFrameRate <= 1000 ){
        cfg->injectorFrameRate = cfg->injectorFrameRate << 16;
    }
later in EbTime.c
double                 injectorInterval  = (double)(1<<16)/(double)injectorFrameRate;

The injectorFrameRate and other APIs are for SvtHevcEncApp. The other implementation(ffmpeg, gst plugin) which use raw svt lib has to handle the api the same way with SvtHevcEncApp.

**Comment by tianjunwork** [comment]
@Yeongjin-Jeong I appreciate your closed attention:) I will address the review comments. 
For ("pred-struct", "deblocking", "sao", "constrained-intra"), they are less frequently changed APIs. You can have them now for add them later on if community requests. Your choice:)

File: gstreamer-plugin/gstsvthevcenc.c (hunk 2 above)
Comment (Yeongjin-Jeong) [reply gstreamer-plugin/gstsvthevcenc.c]: I understand that config param in document (e.g. svt-hevc_encoder_user_guide) is for APP layer :)

**Comment by Yeongjin-Jeong** [comment]
> @Yeongjin-Jeong I appreciate your closed attention:) I will address the review comments.
> For ("pred-struct", "deblocking", "sao", "constrained-intra"), they are less frequently changed APIs. You can have them now for add them later on if community requests. Your choice:)

Okay, If there is any possibility of a change, I will add them later :)

File: gstreamer-plugin/gstsvthevcenc.c [hunk 5]
```diff
@@ -132,9 +134,9 @@ enum
 #define PROP_TILE_COL_DEFAULT               1
 
 #if G_BYTE_ORDER == G_LITTLE_ENDIAN
-#define FORMATS "I420, Y444, I420_10LE, Y444_10LE"
+#define FORMATS "I420, Y444, I420_10LE, I422_10LE, Y444_10LE"
```
Comment (Yeongjin-Jeong) [inline gstreamer-plugin/gstsvthevcenc.c]: Should not you also add Y42B? It means planar 4:2:2 YUV format.

File: gstreamer-plugin/gstsvthevcenc.c [hunk 6]
```diff
@@ -159,7 +158,7 @@ GST_STATIC_PAD_TEMPLATE ("src",
         "width = (int) [ 64, 8192 ], " "height = (int) [ 64, 4320 ], "
         "stream-format = (string) byte-stream, "
         "alignment = (string) au, "
-        "profile = (string) { main, main-10, main-4:4:4}")
+        "profile = (string) { main, main-10, main-4:4:4, main-4:4:4-10 }")
```
Comment (tianjunwork) [inline gstreamer-plugin/gstsvthevcenc.c]: Hi, @Yeongjin-Jeong, I also added main-444-10 in the profile. and the logic to negotiate.

**Pull request merged**